crates/runtime/src/memory.rs

```diff
@@ -130,11 +139,14 @@ impl RuntimeLinearMemory for MmapMemory {
 
     fn grow_to(&mut self, new_size: usize) -> Option<()> {
         if new_size > self.mmap.len() - self.offset_guard_size - self.pre_guard_size {
-            // If the new size is within the declared maximum, but needs more memory than we
-            // have on hand, it's a dynamic heap and it can move.
+            // If the new size of this heap exceeds the current size of the
+            // allocation we have, then this must be a dynamic heap. Use
+            // `new_heap` to calculate a new size of an allocation, allocate it,
+            // and then copy over the memory from before.
             let request_bytes = self
                 .pre_guard_size
                 .checked_add(new_size)?
+                .checked_add(self.extra_to_reserve_on_growth)?
```

Comment:
Would it make sense to consider an exponential growth scheme (like `Vec`, continuing your analogy)? Something simple like doubling (or a factor of `1.5` or pick your favorite value) would give us amortized-constant growth cost.

Reply:
Also, I note a reference to `new_heap` above but I don't see it anywhere -- maybe this was meant to be a helper to adjust the request size somehow?

Reply:
I was wondering if I should have some sort of fancier selection of the growth algorithm here, but I ended up deciding against that and going with something simple for now, but likely to work for awhile since the default reservation size is 2gb. 

I fear I don't personally know enough about how memory64 is expected to be used to know what the best option here is. I'd be afraid with a factor-based growth that you're always reserving 2x the virtual memory which may be a problem for some users. I'm semi-hoping that with a constant amount this is good enough to fly under some radars but immediately bad enough to get flagged by other radars to get a better solution here.

That being said I am also sort of primarily situation-solving a fuzz-bug found where we timed out in memory growth here. I'm *hoping* that the eventual ask is that even for 64-bit memories we still want "static" heaps where the static size is limited to >=4gb. For example we'd reserve 8gb and never move it but we wouldn't grow beyond 8gb (or something like that). In that scenario we wouldn't even hit the dynamic heap path since the maximum size would prove that unnecessary.

Overall how do you feel about sitting on a simple add-the-constant-in strategy for now? I think it may be best to write some of this up in the `Config` documentation, basically saying "if this doesn't work for you please let us know and we can figure out a better solution"?

Reply:
Sure, a simple constant-growth scheme could be fine for now, if we're unsure about usage patterns and resource limits; we can always refine later.

Re: fuzzing specifically, I can foresee the fuzzer working out that there is a fixed increment beyond which the growth occurs, and then just incrementing by that amount each time in order to find another path to the timeout. This concern goes away if the overall limit imposed by fuzzing (at the process VM limit level or within wasmtime proper) is only a small multiple of the extra growth space; e.g. if we grow with an extra 2GB and the limit during fuzzing is 8GB then we can only grow four times, so that's fine.

Zooming out a bit, I do agree that setting a max growth limit and reserving that much VM space upfront is more likely the reasonable approach in the long run; it's hard for me to think of a heap-size limit that is large enough to be plausibly used in some use-case (very large movie-editing software, or in-memory cache, or...) at a few hundred GB or so, yet also too large to just reserve the address space upfront. But let's cross that bridge when we get there!

Reply:
Ok yeah I think I'll document our concerns here and probably go with this. I actually chose the 2gb default because it's bigger than our fuzz limit of 1gb so the fuzzers are guaranteed to never find this corner case. Or at least *our* fuzzers won't find it... But I agree that it's just pushing the problem back, not actually solving anything as of yet.